ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -544,4 +528,148 @@ private boolean containsName(String name) {
+        logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
+        accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+        throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
+      }
+      maybePersistOldState(accountMap);
+
+      // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+      // be caught by Helix and helixStore#update will return false.
+      if (hasConflictingAccount(accountsToUpdate, remoteAccountInfoMap)) {
+        // Throw exception, so that helixStore can capture and terminate the update operation
+        throw new IllegalArgumentException(
+            "Updating accounts failed because one account to update conflicts with existing accounts");
+      } else {
+        for (Account account : accountsToUpdate) {
+          try {
+            accountMap.put(String.valueOf(account.getId()), account.toJson(true).toString());
+          } catch (Exception e) {
+            String message = "Updating accounts failed because unexpected exception occurred when updating accountId="
+                + account.getId() + " accountName=" + account.getName();
+            // Do not depend on Helix to log, so log the error message here.
+            logger.error(message, e);
+            throw new IllegalStateException(message, e);
+          }
+        }
+        newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
+        potentialNewState = accountMap;
+        return newRecord;
+      }
+    }
+
+    /**
+     * Save the zookeeper state from after the update to disk. This will only save the file if the update succeeded
+     * and the old state backup file was successfully reserved.
+     * The following file name format will be used: {@code {yyyyMMdd}T{HHmmss}.{unique long}.new}.
+     * @param succeeded {@code true} iff the update succeeded.
+     */
+    void maybePersistNewState(boolean succeeded) {
+      if (backupPrefixAndPath != null && succeeded) {
+        try {
+          Path filepath = backupDirPath.resolve(backupPrefixAndPath.getFirst() + NEW_STATE_SUFFIX);
+          writeBackup(filepath, potentialNewState);
+        } catch (Exception e) {
+          logger.error("Could not write new state backup file", e);
+          accountServiceMetrics.backupErrorCount.inc();
+        }
+      }
+    }
+
+    /**
+     * Save the zookeeper state from before the update to disk.
+     * The following file name format will be used: {@code {yyyyMMdd}T{HHmmss}.{unique long}.old}.
+     * If there are multiple files with the same timestamp, the unique long will prevent the file names from clashing.
+     */
+    private void maybePersistOldState(Map<String, String> oldState) {
+      if (backupDirPath != null) {
+        try {
+          if (backupPrefixAndPath == null) {
+            backupPrefixAndPath = reserveBackupFile();
```

Comment:
minor: why not do this in the constructor?

Reply:
moved to constructor